Scrubsanity

garrettjoecox opened the pull request:
TODO:
- [ ] Moar testing
- [x] Add unique message for 0 rupees
- [x] Wait for  #1191  to be merged in
- [x] Ensure this doesn't break vanilla
- [x] Ensure this doesn't break scrubs without scrubsanity enabled (probably does right now)
- [x] Scrubs won't sell you items twice, but they will always be spawned, need to prevent them from spawning
- [x] Support Random & Expensive scrubsanity settings
- [x] Todo comment on `z_en_dns.c` Line 413
- [x] OG Item argument for randomizer call is hardcoded to GI_STICKS_1, should be whatever the vanilla scrub offers

garrettjoecox commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1707,8 +1701,8 @@ extern "C" int CustomMessage_RetrieveIfExists(GlobalContext* globalCtx) {
             }
-        } else if (textId == TEXT_SCRUB_POH || textId == TEXT_SCRUB_STICK_UPGRADE || textId == TEXT_SCRUB_NUT_UPGRADE) {
-            messageEntry = Randomizer_GetScrubMessage(textId);
+        } else if (textId >= 0x9000 || textId <= 0x905F) {
```
This was my solution to getting around not having access to actorParams, this is also what 3drando uses for scrub messages.

leggettc18 replied:
A couple of points here. 3ds and n64 text stuff doesn't necessarily match up from my experience, so we need to make sure those are actually the ones the game uses. This might be what's causing the bug Flax found with the recovery hearts, for instance.

leggettc18 replied:
This should be `&&`, not `||`. With `||`, all textIds below `0x905F` were matching, and I think a few items were matching textIds in our Scrub Message Table.

garrettjoecox replied:
```suggestion
        } else if (textId >= 0x9000 && textId <= 0x905F) {
```

garrettjoecox committed:
Initial work towards scrubsanity, got affordable prices mostly functional

Write and read prices from spoiler, move IdentifyScrub up to Randomizer

In a semi working state

Fix bug when going from c to cpp

Kill scrubs before they spawn if already purchased

Woopsie from rebase

Update soh/soh/OTRGlobals.cpp

garrettjoecox committed:
Merge branch 'rando-next' of https://github.com/HarbourMasters/Shipwright into scrubsanity

garrettjoecox committed:
Custom message for free items

garrettjoecox committed:
Merge branch 'rando-next' of https://github.com/HarbourMasters/Shipwright into scrubsanity

aMannus commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1592,11 +1592,6 @@ extern "C" CustomMessageEntry Randomizer_GetScrubMessage(u16 scrubTextId) {
     return CustomMessageManager::Instance->RetrieveMessage(Randomizer::scrubMessageTableID, price);
 }
 
-extern "C" CustomMessageEntry Randomizer_GetNaviMessage() {
-    u16 naviTextId = rand() % NUM_NAVI_MESSAGES;
-    return CustomMessageManager::Instance->RetrieveMessage(Randomizer::NaviRandoMessageTableID, naviTextId);
-}
-
```
Was this removed during a merge by accident?

garrettjoecox replied:
Indeed, this is causing the failure.

aMannus commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3130,11 +3414,20 @@ void DrawRandoEditor(bool& open) {
-                    SohImGui::EnhancementCheckbox(Settings::ShuffleCows.GetName().c_str(), "gRandomizeShuffleCows");
-                    InsertHelpHoverText(
-                        "Cows give a randomized item from the pool upon performing Epona's Song in front of them.");
-                    PaddedSeparator();
+                        // Shuffle Scrubs
+                        ImGui::Text(Settings::Scrubsanity.GetName().c_str());
+                        InsertHelpHoverText(
+                            "Off - Scrubs will not be shuffled.\n"
+                            "\n"
+                            "Affordable - Scrubs will be shuffled and their item will cost 10 rupees.\n"
+                        );
+                        SohImGui::EnhancementCombobox("gRandomizeShuffleScrubs", randoShuffleScrubs, 4, 0);
+                        PaddedSeparator();
+
+                        // Shuffle Cows
+                        SohImGui::EnhancementCheckbox(Settings::ShuffleCows.GetName().c_str(), "gRandomizeShuffleCows");
+                        InsertHelpHoverText("Cows give a randomized item from the pool upon performing Epona's Song in front of them.");
+                        PaddedSeparator();
 
```
Indentation is off, and it seems like you're missing text in the tooltip for the "Expensive" and "Random Prices" options. I assume you made the tooltip before you made the other options possible.

Also, feel free to move the scrubsanity dropdown above tokensanity. That way we have all the dropdowns grouped up together.

aMannus commented on soh/soh/Enhancements/randomizer/randomizerTypes.h:
```diff
@@ -1015,3 +1017,11 @@ typedef enum {
+    GetItemID getItemId;
+    int32_t itemPrice;
+    bool isShuffled;
+} ScrubIdentity;
```
Missing newline ;-)

aMannus reviewed (commented):
Couple of comments from me. Looks great though!

garrettjoecox committed:
Fix issue from bad merge and a few other small things

briaguya0 commented on soh/soh/Enhancements/randomizer/3drando/fill.cpp:
```diff
@@ -1039,6 +1039,19 @@ int Fill() {
     std::vector<uint32_t> remainingPool = FilterAndEraseFromPool(ItemPool, [](const auto i) { return true; });
     FastFill(remainingPool, GetAllEmptyLocations(), false);
+
+    // Add prices for scrubsanity
```
might be nice to elaborate on why we need to do this for soh (since it's not happening in 3drando), even just replacing the current comment with
```cpp
// in soh, we read scrub item prices from the spoilerfile
```
would probably do the trick imo

briaguya0 commented on soh/soh/Enhancements/randomizer/3drando/item_location.cpp:
```diff
@@ -1011,6 +1011,56 @@ std::vector<std::vector<uint32_t>> ShopLocationLists = {
 };
 
+//List of scrubs, used for pricing the scrubs
+std::vector<uint32_t> ScrubLocations = {
```
i kind of wish we didn't have to manually list these here, but it does seem like the most efficient way to handle it at runtime

briaguya0 commented on soh/soh/Enhancements/randomizer/3drando/spoiler_log.cpp:
```diff
@@ -673,15 +673,14 @@ static void WriteHints(int language) {
-            case 2:
-                jsonData["locations"][location->GetName()] = location->GetPlacedItemName().french;
-                break;
+        std::string placedItemName = language == 2 ? location->GetPlacedItemName().french : location->GetPlacedItemName().english;
```
i think the idea here is we will eventually want to support german too, so using the switch case was forward thinking

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -1587,6 +1602,280 @@ std::string Randomizer::GetGanonHintText() const {
+    scrubIdentity.isShuffled = GetRandoSettingValue(RSK_SHUFFLE_SCRUBS) > 0;
+
+    switch (actorParams) {
+        case 0x00:
```
any chance we could make an enum with these values so it's not magic numbery? (i know we have other switch cases full of magic numbers in randomizer.cpp but it'd be nice to move away from that

briaguya0 replied:
actually, it's not clear to me what the purpose of this part is. is this for having items for after we've gotten the "mysterious" item?

briaguya0 replied:
coming back to this after seeing how it's used, seems it's an ogID thing

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -1587,6 +1602,280 @@ std::string Randomizer::GetGanonHintText() const {
+        case SCENE_DDAN: // Dodongo's Cavern
+            switch (actorParams) {
+                case 0x00:
+                    scrubIdentity.scrubId = 0x00;
```
could be nice to enum these too

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3130,11 +3414,20 @@ void DrawRandoEditor(bool& open) {
+                            "Off - Scrubs will not be shuffled.\n"
+                            "\n"
+                            "Affordable - Scrubs will be shuffled and their item will cost 10 rupees.\n"
+                        );
```
looks like this is missing a few options

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3721,8 +4015,14 @@ void CreateGetItemMessages(std::vector<GetItemMessage> messageEntries) {
+            "\x12\x38\x82\D'accord! Vous avez gagné! En échange&de m'épargner, je vous donnerai un &%gobjet mystérieux%w! S'il vous plaît,&prenez-le!\x07\x10\xA3",
+        });
+
+    for (u32 price = 5; price <= 95; price += 5) {
```
i get that you're making new messages for every price, and then pulling it out using
```cpp
messageEntry = Randomizer_GetScrubMessage((textId & ((1 << 8) - 1)));
```
it just feels a little strange to me

@leggettc18 if you have any thoughts on this use of the custom message system i'd love to hear them

leggettc18 replied:
Well at the moment the price is the only thing that's different between all of these messages, so it made sense to me to use the price as a textId. Since then there's been talk of adding an option for the text to reference the item they sell with differing levels of obscurity (pretty much exactly like hints do currently) which would require either saving the text to the save file (and saving the price along with it) or substituting that item text at runtime (at which point we might as well do that for the price as well). So maybe it should be changed in the future, but this might be fine for the first pass.

garrettjoecox replied:
> or substituting that item text at runtime

This is the pattern I'm following for shopsanity, and plan to eventually migrate scrubs to, I got it most of the way working with scrubs but ran into a snag that I didn't want to hold up scrubsanity for.

briaguya0 replied:
yeah, i'm cool with it as is too. i think having the bitwise logic in the calls is what was throwing me for a loop when trying to read through it the first time

briaguya0 replied:
> This is the pattern I'm following for shopsanity, and plan to eventually migrate scrubs to, I got it most of the way working with scrubs but ran into a snag that I didn't want to hold up scrubsanity for.

good deal! we can make an issue for that to look into after getting this merged

leggettc18 replied:
I think changing the scrubs table to a merchants table would be a good idea. Shopsanity messages can also go in there, and when we do shuffle magic beans and shuffle merchants the changed dialog to those can also go in there.

garrettjoecox replied:
Going to resolve this comment, but added a comment here explaining the plans

briaguya0 commented on soh/soh/OTRGlobals.cpp:
```diff
@@ -1718,8 +1712,8 @@ extern "C" int CustomMessage_RetrieveIfExists(GlobalContext* globalCtx) {
-        } else if (textId == TEXT_SCRUB_POH || textId == TEXT_SCRUB_STICK_UPGRADE || textId == TEXT_SCRUB_NUT_UPGRADE) {
-            messageEntry = Randomizer_GetScrubMessage(textId);
+        } else if (textId >= 0x9000 && textId <= 0x905F) {
+            messageEntry = Randomizer_GetScrubMessage((textId & ((1 << 8) - 1)));
```
i think the `(textId & ((1 << 8) - 1))` logic should move into `GetScrubMessage`, so the `itemPrice` param name makes sense

briaguya0 replied:
looking at this again, since we only have the textId at this point, it's probably worth just passing in the textId to `GetScrubMessage` and then having commented logic for getting the price out of that textId inside `GetScrubMessage`

briaguya0 replied:
seeing
```c
// Currently the textID is simply identified by the item price since that is the only thing
// unique to it, later on this will change to identifying by scrubIdentity.randomizerInf
```
makes me think this can be a later problem

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.h:
```diff
@@ -19,6 +19,7 @@ class Randomizer {
     std::string ganonHintText;
     std::string ganonText;
     std::unordered_map<RandomizerSettingKey, u8> randoSettings;
+    std::unordered_map<RandomizerCheck, u16> scrubPrices;
```
general naming thing, maybe `scrubItemPrices`?

garrettjoecox replied:
Renamed to randomizerMerchantPrices, will re-use for shopsanity

briaguya0 commented on soh/include/z64save.h:
```diff
@@ -185,6 +185,7 @@ typedef struct {
     u8 dungeonsDone[8];
     u8 trialsDone[6];
     u8 cowsMilked[10];
+    u8 scrubsPurchased[35];
```
naming, maybe `scrubItemsPurchased`?

briaguya0 commented on soh/soh/SaveManager.cpp:
```diff
@@ -768,6 +768,10 @@ void SaveManager::LoadBaseVersion1() {
     SaveManager::Instance->LoadArray("cowsMilked", ARRAY_COUNT(gSaveContext.cowsMilked), [](size_t i) {
         SaveManager::Instance->LoadData("", gSaveContext.cowsMilked[i]);
     });
+
+    SaveManager::Instance->LoadArray("scrubsPurchased", ARRAY_COUNT(gSaveContext.scrubsPurchased), [](size_t i) {
+        SaveManager::Instance->LoadData("", gSaveContext.scrubsPurchased[i]);
+    });
```
might need to look into how stuff works with the v1/v2 stuff here (not for this PR since it seems we have cows in both too), but it'll probably come up in the rando v2 review. maybe worth making an issue for it

briaguya0 commented on soh/src/overlays/actors/ovl_En_Dns/z_en_dns.c:
```diff
@@ -164,7 +165,25 @@ void EnDns_Init(Actor* thisx, GlobalContext* globalCtx) {
+    this->actor.textId = D_809F040C[this->actor.params];
     this->dnsItemEntry = sItemEntries[this->actor.params];
+    if (gSaveContext.n64ddFlag) {
+        // Ugly, but the best way we can identify which grotto we are in, same method 3DRando uses, but we'll need to account for entrance rando
```
i know i did something different when identifying grottos for gossip stone hints in https://github.com/briaguya-ai/Shipwright/pull/106/files

any thoughts on how to make sure this works for entrance rando?

garrettjoecox replied:
The method used for grottos is grabbing the chest actor params, this won't work here as there isn't a chest in any of the scrub grottos.

For entrance randomizer we'll just need to make sure that the respawn data is set correctly when entering a scrub grotto. The respawn data will have to be set regardless for entrance rando, if not the grotto exits would be broken, it's just a matter of ensuring it's matched up with what's in the spoiler.

briaguya0 commented on soh/src/overlays/actors/ovl_En_Dns/z_en_dns.c:
```diff
@@ -185,6 +204,13 @@ void EnDns_ChangeAnim(EnDns* this, u8 index) {
+    if (gSaveContext.rupees < this->dnsItemEntry->itemPrice || gSaveContext.scrubsPurchased[this->scrubIdentity.scrubId] == 1) {
+        return 0;
+    }
+    return 4;
```
assuming this is `4` to be `0b100` for some bitwise stuff in the purchaseableCheck

garrettjoecox replied:
It's used here, with 0,1, and 3 being failure states. Probably could be an enum but I'm not sure what the detailed difference between them all is
<img width="342" alt="Screen Shot 2022-08-25 at 1 36 49 PM" src="https://user-images.githubusercontent.com/7316699/186742780-33ee244b-a442-428f-bd94-91f1445fd490.png">

briaguya0 replied:
![image](https://user-images.githubusercontent.com/70942617/186760257-eba61fa7-927d-48cf-9a88-83b080427160.png)
![image](https://user-images.githubusercontent.com/70942617/186760332-1258b269-8947-4bad-bd19-9c15fe16e0e2.png)
![image](https://user-images.githubusercontent.com/70942617/186760382-1c591fee-bc39-458a-8ef5-e3bafe7b4fce.png)

oddest thing in that code is it seems both `2` and `4` get us to the successful purchase.

briaguya0 replied:
anyways, feels like a magic number but it's cool for now

briaguya0 commented on soh/src/overlays/actors/ovl_En_Dns/z_en_dns.c:
```diff
@@ -489,6 +513,9 @@ void EnDns_Update(Actor* thisx, GlobalContext* globalCtx) {
     this->dustTimer++;
     this->actor.textId = D_809F040C[this->actor.params];
+    if (gSaveContext.n64ddFlag && this->scrubIdentity.isShuffled) {
+        this->actor.textId = 0x9000 + this->dnsItemEntry->itemPrice;
```
probably worth throwing a comment in here explaining what's happening with the message table being full of messages for each price

briaguya0 commented on soh/src/overlays/actors/ovl_En_Shopnuts/z_en_shopnuts.c:
```diff
@@ -69,6 +69,15 @@ void EnShopnuts_Init(Actor* thisx, GlobalContext* globalCtx) {
+        ScrubIdentity scrubIdentity = Randomizer_IdentifyScrub(globalCtx->sceneNum, this->actor.params, respawnData);
+
+        if (scrubIdentity.isShuffled && gSaveContext.scrubsPurchased[scrubIdentity.scrubId] == 1) {
+            Actor_Kill(&this->actor);
```
do we want to actor_kill here? how does 64rando handle this? i'd think we'd want to have it so after getting the mysterious item you could get the vanilla item from the scrub but i'm not sure

leggettc18 replied:
N64 makes them disappear. I feel like it'd be more confusing to have them there after you got their item. Also consistent with Shopsanity where the major items are sold out rather than being replaced with vanilla shop items.

briaguya0 replied:
coolcool, sounds like actor_kill works then

briaguya0 reviewed (changes requested):
very cool! left some notes about naming/ways to make stuff a little easier to read. also left some comments with questions. this looks like it's really close to ready though!

garrettjoecox committed:
Merge branch 'rando-next' of https://github.com/HarbourMasters/Shipwright into scrubsanity

garrettjoecox committed:
Various changes from feedback, and use new randomizerInf table

garrettjoecox committed:
Fix issue in spoiler generation

briaguya0 commented on soh/include/z64save.h:
```diff
@@ -184,7 +184,6 @@ typedef struct {
     char ganonText[250];
     u8 seedIcons[5];
     u16 randomizerInf[2];
```
this is no longer big enough

garrettjoecox replied:
Great catch, I made a mental note to do this and forgot :<

briaguya0 reviewed (changes requested):
just the one thing then this can be merged

garrettjoecox committed:
Increase size of randomizerInf
